torch/csrc/autograd/input_metadata.h

```diff
@@ -67,9 +105,26 @@ struct InputMetadata {
     return at::zeros(shape_, options_);
   }
 
+  at::IntArrayRef compute_shape(const at::Tensor& t){
+    if(t.is_nested()){
+      return at::IntArrayRef{};
+    }
+    return t.sizes();
+  }
+
+  const at::Tensor compute_nested_shape_from_tensor(const at::Tensor& t){
+    if(t.is_nested()){
+      auto nested_impl = at::native::get_nested_tensor_impl(t);
+      return nested_impl->get_nested_size_tensor();
+    }
+    return at::Tensor{};
+  }
+
 private:
   const at::TensorOptions options_;
   at::DimVector shape_;
+  bool is_nested_ = false;
+  const at::Tensor nested_shape_ = at::Tensor{};
```

Comment:
One option here is to make this an optional Tensor, i.e. `c10::optional<at::Tensor>`. That can then also double as an indication as to whether this represents InputMetadata of a NestedTensor or not.